earth2mip/lagged_ensembles/core.py

```diff
@@ -138,26 +124,107 @@ async def yield_lagged_ensembles(
                             torch.cuda.synchronize()
                         yield k, ensemble.pop(k), await observations[jj]
 
-        for _ in range(world_size):
-            obs_buffer.popleft()
+    assert not ensemble, len(ensemble)
 
-    assert not ensemble, len(ensemble)  # noqa
 
+def _convert_ensemble_to_cpu_async(
+    buffers: dict[int, torch.Tensor]
+) -> dict[int, torch.Tensor]:
+    """Maybe convert gpu ensemble to cpu ensemble asynchronously
 
-def num(n, ell, j, L):
-    a = max(ell - j, ell - n, -L)
-    b = min(ell, L)
-    return b - a + 1
+    Before using the output be sure to synchronize the cuda stream using either
+
+        torch.cuda.syncrhonize()
+
+    or cuda Events.
+
+    """
+    if not buffers:
+        return {}
+
+    first_buffer = next(iter(buffers.values()))
+    if first_buffer != torch.device("cpu"):
+        cpu_buffers = {
+            r: torch.empty_like(b, device="cpu", pin_memory=True)
+            for r, b in buffers.items()
+        }
+        for r in buffers:
+            cpu = cpu_buffers[r]
+            gpu = buffers[r]
+            cpu.copy_(gpu, non_blocking=True)
+    else:
+        cpu_buffers = buffers
+    return cpu_buffers
 
 
-def done(j, ell, i, lags, n):
-    """Unused helper function wich says if lag ell and valid_time j are written
-    to in a given iteration `i` of the loop in lagged_average_simple
+def scatter(rank, world_size, i0, nt, y, recv_buffer, lower, upper):
+    """
+
+    constraints:
+        ii = i0 * world_size + src
+        jj = ii + lead_time
+
+        i = jj - lead_time - m
+        i % world_size = dst:
+        lower <= m <= upper
+
+        dst = i0 * world_size + src + lead_time - lead_time - m
+
+        dst = (src - m) % world_size
+        src = (dst + m) % world_size
+
+    """
+    if torch.distributed.is_initialized():
+        buffers = {}
+
+        # too avoid redundant comms when the # of lags > world_size
+        # upper - lower + 1  <= world_size
+        # upper <= world_size + lower - 1
+        upper = min(upper, world_size + lower - 1)
+        for src in range(world_size):
+            src_initial_time = i0 * world_size + src
+            if src_initial_time >= nt:
+                continue
+
+            for m in range(lower, upper + 1):
+                dst = (src - m) % world_size
+                if rank == src == dst:
+                    buffers[rank] = y
+                elif rank == src:
+                    torch.distributed.send(y, dst=dst)
+                elif rank == dst:
+                    torch.distributed.recv(recv_buffer, src=src)
+                    buffers[src] = recv_buffer.clone()
+    else:
+        buffers = {rank: y}
 
-    This is one way to implement the done criteria which is less easily
-    parallelized. I am leaving it in the code for educational value only.
+    return buffers  # noqa
+
```

Comment:
This code seems to be misplaced in this file. A lot of the functionality here does not seem to rely on the assumption of `lagged ensemble`. The design principle should be to move these utilities elsewhere, in a distributed utliity focused area (maybe even modulus.distributed) and to generalize sufficiently so that we can reuse code if necessary.

Reply:
Could you be more specific about what code you see as general purpose here?

Reply:
1. `_convert_ensemble_to_cpu_async` seems like a general purpose utility, not specific to lagged ensemble.
2. `scatter` also seems to be more general purpose utility. As a side note, if this function is meant to be private then we should mark it so, otherwise it would be helpful to have documentation on the inputs.

Reply:
Let’s defer this. This pr is not defining new general purpose utilities for scattering an ensemble distributed across ranks. Just moving code into helpers as appropriate for improved legibility or the lagged ensemble code.

Reply:
That's fine, we can resolve with a link to an issue. Perhaps the issue can be broadly centered on distributed utilities.

Reply:
on more thought. `scatter` is specific to the distribution pattern used in this particular script, but I can open an issue for the `cpu` async.

Reply:
https://github.com/NVIDIA/earth2mip/issues/160